### PR TITLE
mktempdir() now supports prefix.

### LINKED ISSUE #22922: mktempdir() should provide an option of a prefix very similar to you can do with mkdtemp in posix
I think it may help if `mktempdir()` can provide an option of a prefix very similar to you can do with `mkdtemp` in POSIX. This provides the module developers to create temp directories with their own prefix which can help them able to track the directories created by them than everything starting with `tmp` or `jl_` prefix.

The discussion thread on this topic can be seen here: https://discourse.julialang.org/t/mktempdir-should-provide-scope-to-declare-a-prefix/5020

### PR DESCRIPTION
**Pull request by sambitdash**
The libuv is used for this implementation in a platform independent manner. This is for the issue #22922. The `mktemp()` or `tempname()` functions are not affected by this change. 

This may be still acceptable as a module owner can create her own temp directory and then create files inside that. Hence, the importance of prefix on files may not be as significant. 

Edit: fix #22922

### CODE REVIEW

**Comment by sambitdash** [comment]
With this #22998 can be closed.

**Comment by sambitdash** [comment]
Added test cases which show behavioral differences across OS. For example, '*' is not an accepted prefix in Windows which is perfectly permitted in Linux as part of a filename.

**Comment by sambitdash** [comment]
Error reported in Mac OS X is completely unrelated. It's an InterruptException which this code does not deal with. The Windows 32-bit case is more like incomplete run and is not a true failure.

**Comment by sambitdash** [comment]
Closing as there is not much of activity in past week on this. If you there is interest in this change let me know, I will be happy to reopen.

**Comment by StefanKarpinski** [comment]
Seems a bit premature to close this. @nalimilan, you seemed engaged on the original issue here. Can you review this and see if it addresses the problem? Frankly, I don't have enough context to understand what problem this is solving, but that doesn't mean it's not a good idea.

**Comment by sambitdash** [comment]
PR#23259  affected the last test errors. Now fixed.

**Comment by sambitdash** [comment]
Travis failure seems unrelated.

**Comment by StefanKarpinski** [comment]
@vtjnash has said he can take a look at this in a few days. Sorry for the delay, but this is a useful change.

File: base/file.jl [hunk 1]
```diff
@@ -356,12 +339,33 @@ is an open file object for this path.
...
-If `parent` does not exist, throw an error.
+If `parent` does not exist, throw an error. An optional `prefix` to the directory name can
+be provided.
 """
-mktempdir(parent)
+function mktempdir(parent=tempdir(); prefix="$temp_prefix")
+    template = prefix*"XXXXXX"
+    tpath = joinpath(parent, template)
+
+    req = Libc.malloc(_sizeof_uv_fs)
```
Comment (nalimilan) [inline base/file.jl]: I see you use the same approach as for `readlink`, but `readdir` uses a different strategy, by using `zeros`, which allows skipping the `try... finally`. @vtjnash Which approach is recommended?

File: test/file.jl [hunk 2]
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
...
+        filename = basename(tmpdir)
+        @test startswith(filename, tst_prefix)
+    end
+
+    # Behavioral differences across OS types
+    if Sys.iswindows()
+        @test_throws Base.UVError mktempdir(; prefix="*")
+        @test_throws Base.UVError mktempdir(; prefix="cdcdccd/")
+    else
+        # '/' is accepted in a prefix but affects the overall path and permissions.
```
Comment (nalimilan) [inline test/file.jl]: That comment sounds misleading, since `/` is not accepted by the next line.

File: test/file.jl [hunk 3]
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
...
+    tst_prefix="#!@%^&()"
+    mktempdir(; prefix=tst_prefix) do tmpdir
+        filename = basename(tmpdir)
+        @test startswith(filename, tst_prefix)
+    end
+
+    # Behavioral differences across OS types
+    if Sys.iswindows()
+        @test_throws Base.UVError mktempdir(; prefix="*")
+        @test_throws Base.UVError mktempdir(; prefix="cdcdccd/")
```
Comment (nalimilan) [inline test/file.jl]: Does `/` without any characters before it also throw? What about `\` or `C:\`?

File: test/file.jl [hunk 4]
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
...
+
+        # The file created will be of format "/tmp/XXXXXX"
+        mktempdir("/"; prefix="tmp/") do tmpdir
+            filename = basename(tmpdir)
+            @test length(filename) == 6
+        end
+    end
+
+    # Unicode test
+    tst_prefix="\u2200x\u2203y"
```
Comment (nalimilan) [inline test/file.jl]: Directly use the Unicode chars, that's clearer.

**Review by nalimilan — COMMENTED** [review]
Thanks! Looks mostly good, though I'd like a deeper review of libuv parts by somebody familiar with that code. Please improve the commit message to explain that you switch to the new `uv_fs_mkdtemp` function, and which libuv version it was introduced in.

File: base/file.jl [hunk 5]
```diff
@@ -356,12 +339,33 @@ is an open file object for this path.
...
+    try
+        ret = ccall(:uv_fs_mkdtemp, Int32,
+            (Ptr{Void}, Ptr{Void}, Cstring, Ptr{Void}),
```
Comment (sambitdash) [inline base/file.jl]: I followed the same indentation as `readlink`, although my personal preference would be to align to ccall.

File: base/file.jl (hunk 1 above)
Comment (sambitdash) [inline base/file.jl]: I guess I now get the approach here. In `jl_uv.c` all request structures are allocated in stack (declared as local var) and passed on to the function in the c-layer. The cleanup is carried out on the address to the request like &req. In the Julia layer you cannot have a stack variable so `zero` is a means to create a Julia UInt8 array which will me memory managed in the Julia layer. Hence, no special action on a throw and clean up will be called on this object after the request object is used in the `ccall`. The approach is a bit hard to understand initially.  Are `try`...`catch` overheads very significant? I can change the code to the `zero` approach.

File: test/file.jl (hunk 2 above)
Comment (sambitdash) [inline test/file.jl]: The whole behavior is very messy. Here are the full set of samples. I think we should just not accept '/' in prefix. Some samples:

```
julia> mktempdir("/";prefix="tmp/")
"/tmp/m12rZm"

julia> mktempdir("/tmp";prefix="/")
ERROR: mktempdir: permission denied (EACCES)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] #mktempdir#12(::String, ::Function, ::String) at ./file.jl:359
 [3] (::Base.Filesystem.#kw##mktempdir)(::Array{Any,1}, ::Base.Filesystem.#mktempdir, ::String) at ./<missing>:0
```

Comment (nalimilan) [inline test/file.jl]: I'm not too worried about this, as these behaviors make sense, but differences between OSes are annoying. What happens with `\` on Windows?

File: base/file.jl (hunk 1 above)
Comment (nalimilan) [inline base/file.jl]: I would think the simplest approach is the best one, but others must know better.

File: test/file.jl (hunk 3 above)
Comment (sambitdash) [inline test/file.jl]: Give me a couple of days. My windows VM is needs some fixing before I can test this. I am a native Linux programmer :-).

**Comment by sambitdash** [comment]
closing and resubmitting with rebasing.

**Comment by fredrikekre** [comment]
You can rebase and force-push to the same branch and keep the same PR.

**Comment by sambitdash** [comment]
@fredrikekre  The forced push detached the PR and could not reopen it. The new PR on this is #23634.

**Comment by fredrikekre** [comment]
Yea, I think you need to reopen before force-pushing in such a case.

**Pull request closed without merging**